server/oidc.go

```diff
@@ -0,0 +1,269 @@
+		return
+	}
+
+	token, err := o.Exchange(
+		context.Background(),
+		c.Query("code"),
+		oauth2.SetAuthURLParam("code_verifier", session.Values["oauth_code_verifier"].(string)),
+	)
+	if err != nil {
+		c.AbortWithError(http.StatusUnauthorized, err)
+		return
+	}
+
+	// Extract the ID Token from OAuth2 token.
+	rawIDToken, ok := token.Extra("id_token").(string)
+	if !ok {
+		c.AbortWithError(http.StatusUnauthorized, fmt.Errorf("id token missing from OAuth2 token"))
+		return
+	}
+	_, err = o.params.verifyToken(rawIDToken)
+
+	if err != nil {
+		c.AbortWithError(http.StatusForbidden, err)
+		return
+	} else {
+		session.Values["access_token"] = token.AccessToken
+
+		session.Save(c.Request, c.Writer)
+	}
+
+	c.Redirect(http.StatusFound, o.clientRedirect)
+}
+
+// HandleOIDCLogin handles redirecting the user to the Okta login, as well as
+// providing it challenge codes.
+// See: https://developer.okta.com/docs/guides/sign-into-web-app-redirect/go/main/
+func (o *oauthEnv) HandleOIDCLogin(c *gin.Context) {
+	c.Header("Cache-Control", "no-cache") // See https://github.com/okta/samples-golang/issues/20
+
+	session, err := sessionStore.Get(c.Request, oktaCookieName)
+	if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, err)
+		return
+	}
+	// Generate a random state parameter for CSRF security
+	oauthState := randstr.Hex(16)
+
+	// Create the PKCE code verifier and code challenge
+	oauthCodeVerifier, err := oauthUtils.GenerateCodeVerifierWithLength(50)
+	if err != nil {
+		c.AbortWithError(http.StatusInternalServerError, err)
+		return
+	}
+	// get sha256 hash of the code verifier
+	oauthCodeChallenge := oauthCodeVerifier.CodeChallengeS256()
+
+	session.Values["oauth_state"] = oauthState
+	session.Values["oauth_code_verifier"] = oauthCodeVerifier.String()
+
+	session.Save(c.Request, c.Writer)
+
+	redirectURI := o.AuthCodeURL(
+		oauthState,
+		oauth2.SetAuthURLParam("code_challenge", oauthCodeChallenge),
+		oauth2.SetAuthURLParam("code_challenge_method", "S256"),
+	)
+
+	c.Redirect(http.StatusFound, redirectURI)
+}
+
+// verifyToken passes the token and the oAuthParameters through a JWT verifier
+// See: https://developer.okta.com/docs/guides/sign-into-web-app-redirect/go/main/
+func (p oAuthParameters) verifyToken(t string) (*verifier.Jwt, error) {
+	tv := map[string]string{}
+	tv["aud"] = p.clientID
+	jv := verifier.JwtVerifier{
+		Issuer:           p.issuer,
+		ClaimsToValidate: tv,
+	}
+
+	result, err := jv.New().VerifyIdToken(t)
+	if err != nil {
+		return nil, err
+	}
+
+	if result != nil {
+		return result, nil
+	}
+
+	return nil, fmt.Errorf("token could not be verified")
+}
+
+// getProfileData takes a HTTP request (containing things like the cookie) and
+// will get user information in a map from Okta
+// See: https://developer.okta.com/docs/guides/sign-into-web-app-redirect/go/main/
+func (s *Server) getProfileData(r *http.Request) (map[string]string, error) {
```

Comment:
I think there's a few distinct things happening in the body of this function, and it would be easier to follow if it was refactored out to multiple funcs.